# Support Vector Machine

Sentiment classification on the IMDB Dataset using SVM to predict positive or negative reviews.

In [1]:
import numpy as np
from tensorflow.keras.datasets import imdb
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score, classification_report

In [2]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=10000)

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, seq in enumerate(sequences):
        results[i, seq] = 1
    return results

x_train_vec = vectorize_sequences(x_train)
x_test_vec = vectorize_sequences(x_test)

17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


In [3]:
svm_model = LinearSVC()
svm_model.fit(x_train_vec, y_train)
y_pred_svm = svm_model.predict(x_test_vec)

In [4]:
print("SVM Accuracy:", accuracy_score(y_test, y_pred_svm))
print(classification_report(y_test, y_pred_svm))

SVM Accuracy: 0.83596
              precision    recall  f1-score   support

           0       0.83      0.84      0.84     12500
           1       0.84      0.83      0.83     12500

    accuracy                           0.84     25000
   macro avg       0.84      0.84      0.84     25000
weighted avg       0.84      0.84      0.84     25000



The SVM model achieves about 83.60% accuracy, with precision, recall, and F1-scores around 0.83 to 0.84 for both classes. The performance is slightly lower than Logistic Regression, but still consistent and balanced across positive and negative reviews.